libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -2229,6 +2230,70 @@ bool WaveTrack::GetOne(samplePtr buffer,
    return result;
 }
 
+std::vector<TrackSampleView> WaveTrack::GetSampleView(
+   size_t iChannel, size_t nBuffers, sampleCount start, size_t length) const
+{
+   const auto nChannels = NChannels();
+   assert(nBuffers - iChannel <= nChannels); // precondition
+   const auto pOwner = GetOwner();
+   if (!pOwner) {
+      // GetSampleView called on a WaveTrack outside a project ??
+      // Not a complication I want to address, all the more that ongoing
+      // refactoring should rid us of this call.
+      assert(false);
+      return {};
+   }
+   std::vector<TrackSampleView> result;
+   auto iter = pOwner->Find<const WaveTrack>(this);
+   for (auto i = 0u; i < nBuffers;++i) {
+      auto pTrack = *iter++;
+      result.push_back(pTrack->GetOneSampleView(start, length));
+   }
+   return result;
+}
+
+TrackSampleView
+WaveTrack::GetOneSampleView(sampleCount start, size_t length) const
+{
+   WaveClipConstHolders intersectingClips;
+   auto t0 = LongSamplesToTime(start);
+   const auto t1 = t0 + static_cast<double>(length) / GetRate();
+   for (const auto& clip : mClips)
+      if (t0 < clip->GetPlayEndTime() && clip->GetPlayStartTime() < t1)
+         intersectingClips.push_back(clip);
+   if (intersectingClips.empty())
+      return {};
+   std::sort(
+      intersectingClips.begin(), intersectingClips.end(),
+      [](const auto& a, const auto& b) {
+         return a->GetPlayStartTime() < b->GetPlayStartTime();
+      });
+   std::vector<AudioSegmentSampleView> segments;
+   segments.reserve(intersectingClips.size());
+   for (auto i = 0u; i < intersectingClips.size();++i)
+   {
+      const auto& clip = intersectingClips[i];
+      const auto clipStartTime = clip->GetPlayStartTime();
+      if (t0 < clipStartTime) {
+         // Insert silence
+         const auto numSamples = TimeToLongSamples(clipStartTime - t0);
+         std::vector<float> silence(numSamples.as_long_long());
+         std::vector<BlockSampleView> pointers = {
+            std::make_shared<std::vector<float>>(std::move(silence))
+         };
+         segments.push_back({ std::move(pointers), 0u, numSamples });
+         t0 = clipStartTime;
+      }
+      const auto clipT0 = t0 - clipStartTime;
+      const auto clipS0 = TimeToLongSamples(clipT0);
+      const auto len = TimeToLongSamples(t1 - clipT0).as_size_t();
```

Comment:
What's a proof that this is a save `as_size_t` ?  I think it can be deduced that `len` is not more than `length`.  An `assert` of that with some explanation of that deduction would be welcome.

Reply:
`as_size_t` already does the assertion, but I'm adding a comment.